### DISCUSSION TITLE
[Question] Using progress component to show download progress

### DISCUSSION THREAD

**Discussion by ghost** [charmbracelet/bubbles#127] [Q&A] [answered]
How to use the progress component to show download progress. I have tried couple of things but in vain. Can someone provide an example?

**Comment by bashbunni** [comment]
zshbunni >:o my long lost relative

**Comment by bashbunni** [comment]
I guess I'll also comment something potentially helpful. One of our developers @caarlos0 has a simple timer tool that uses the `progress.Model` bubble. Since [his example](https://github.com/caarlos0/timer/blob/main/main.go) is quite a simple program, it might help to answer your question. Let us know if that helps or if you'd like more assistance!

**Comment by ghost** [comment]
I have a function called `downloadFile` which uses `net/http` to download
any file given valid URL. Using `io.TeeReader` I am able to get the downloads progress in
bytes. I thought that using this I can hook this up with progress component.

My question is how can I update my bubbletea program model from this function `downloadFile`?

Sorry if this question is not clear. Let me know if you want me to provide any code samples.

**Comment by bashbunni** [comment]
What you'll want to do is have a `tea.Msg` that gets handled in the `Update` method. This `tea.Msg` is returned from the function you use to grab the progress info from `downloadFile`. You might have a function called `getProgress` that returns the percentage of the download completed based on the info from `downloadFile`
When that message is received in your Update method, you can set the percentage of completion, then call the function again to get another update. You can continue this cycle until the download has completed. 

It'll likely end up looking something like this:

```go
﻿type progressMsg float64

func getProgress() tea.Msg {
    f, _ := getProgress()
    return progressMsg{f}
}

func (m model) Update(msg tea.Msg) tea.Model, tea.Cmd {
    switch msg := msg.(type) {
    ...
    case progressMsg:
        if m.progress.Percent() == 1.0 {
            return m, tea.Quit 
            // or whatever other command you want to happen when it has finished downloading
        }
        cmd := m.progress.SetPercent(progressMsg.f) 
        // might need to be progressMsg.f / total depending on what data you're getting back
        return m, tea.Batch(getProgress(), cmd)
    }
}
```
[This](https://github.com/charmbracelet/bubbletea/blob/master/examples/progress-animated/main.go) example might also help for more context

Let me know if this helps. To be more specific, I'd have to see your code

**Comment by ghost** [comment]
I am using `io.TeeReader` to keep track of the bytes downloaded. There's no way to inform about this event to my bubbletea program.

I looked at the example but still my usecase is far different from the one provided but thanks anways

Here's the code:

```go
package main

import (
	"flag"
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"
)

type progressWriter struct {
  total int
}

func (pw *progressWriter) Write(p []byte) (int, error) {
  pw.total += len(p)
  pw.Print()
  return len(p), nil
}

func (pw *progressWriter) Print() {
  fmt.Printf("\rDownloaded %d bytes", pw.total)
}

func downloadFile(url string) error {
  filename := filepath.Base(url)

  file, err := os.Create(filename)
  if err != nil {
    return err
  }

  resp, err := http.Get(url)
  if err != nil {
    return err
  }
  defer resp.Body.Close()

  if resp.StatusCode != 200 {
    return fmt.Errorf("receiving status of %d for url: %s", resp.StatusCode, url)
  }

  pw := &progressWriter{}
  _, err = io.Copy(file, io.TeeReader(resp.Body, pw))

  return err
}

func main() {
  url := flag.String("url", "", "url for the file to download")
  flag.Parse()

  err := downloadFile(*url)
  if err != nil {
    fmt.Println("uh oh: ", err)
    os.Exit(1)
  }

  fmt.Println()
}
```

To run `go run . --url="https://filesamples.com/samples/video/mkv/sample_960x540.mkv"`

**Reply by ghost** [reply]
@meowgorithm do you have any suggestions?

**Comment by bashbunni** [comment]
You can send a `tea.Msg` to your Bubble tea program (`tea.Program`) using [`p.Send(tea.Msg)`](https://pkg.go.dev/github.com/charmbracelet/bubbletea#Program.Send). I can post some example code shortly

**Comment by ghost** [comment]
Thank you @bashbunni. This is what I have been looking for all this time. Just slightly worried about this line in the docs

> This method is currently provisional. The method signature may alter slightly, or it may be removed in a future version of this package.

**Reply by meowgorithm** [reply]
We've tested the feature for a bit and it's here to stay. I'll remove that note in the next release.

**Reply by ghost** [reply]
Glad to hear. Thank you

**Reply by meowgorithm** [reply]
✅ Done https://github.com/charmbracelet/bubbletea/pull/271

**Comment by bashbunni** [comment, marked as answer]
Hey @zshbunni in case this is a blocker for you, I wanted to give you an update. We're adding another example that was built off of your question to our documentation as I think it will help many people. You can see the code for yourself [here](https://github.com/charmbracelet/bubbletea/pull/270). Hope that helps! Let us know if you have any questions

**Comment by ghost** [comment]
Hey @bashbunni, I got the necessary info to do what I wanted. Thanks for you assistance and the example. I have other doubts on how to do certain stuff (like modal for eg) but I think it would be appropriate to post it in seperate discussion 🐰

**Reply by bashbunni** [reply]
Awesome! Never hesitate to reach out. We'd be really curious to see what you build :)
